### PR TITLE
CL/HIER: 2lvl alltoallv

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Adds implementation of 2lvl alltoallv in the CL/HIER. 

## Why ?
Allows splitting a2av among TLs (e.g., TL/CUDA for intranode + TL/UCP for internode)

## How ?
CL/HIER allocates 2 sbgps: NODE and FULL, allocates a storage for counts/displs for 2 tasks and initializes it. Then the 2 tasks are launched concurrently.

### CODE REVIEW

File: src/components/cl/hier/cl_hier.c [hunk 1]
```diff
@@ -38,6 +38,17 @@ static ucc_config_field_t ucc_cl_hier_lib_config_table[] = {
      ucc_offsetof(ucc_cl_hier_lib_config_t, sbgp_tls[UCC_HIER_SBGP_NET]),
      UCC_CONFIG_TYPE_STRING_ARRAY},
 
+    {"FULL_SBGP_TLS", "ucp",
+     "TLS to be used for FULL subgroup.\n"
+     "FULL subgroup contains all processes of the team",
+     ucc_offsetof(ucc_cl_hier_lib_config_t, sbgp_tls[UCC_HIER_SBGP_FULL]),
+     UCC_CONFIG_TYPE_STRING_ARRAY},
+
+    {"A2AV_NODE_THRESH", "0",
```
Comment (Sergei-Lebedev) [inline src/components/cl/hier/cl_hier.c:47]: i think we should keep existing naming scheme from TLs, i.e. COLLNAME_ALGNAME_PARAMNAME
for example ALLTOALLV_SPLIT_NODE_THRESH

**Review by Sergei-Lebedev — COMMENTED** [review]
do you think we need similar algorithm for alltoall?

File: src/components/cl/hier/alltoallv/alltoallv.c [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    if (!SBGP_ENABLED(cl_team, NODE) || !SBGP_ENABLED(cl_team, FULL)) {
+        cl_debug(team->context->lib, "alltoallv requires NODE and FULL sbgps");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    c64 = UCC_COLL_ARGS_COUNT64(&coll_args->args);
+    d64 = UCC_COLL_ARGS_DISPL64(&coll_args->args);
+
+    if ((c64 && !d64) || (!c64 && d64)) {
```
Comment (bureddy) [inline src/components/cl/hier/alltoallv/alltoallv.c:129]: `if (c64 ^ d64)` ?

**Review by bureddy — APPROVED** [review]

**Comment by vspetrov** [comment]
> do you think we need similar algorithm for alltoall?

I think we can add if see a use case : does FB run pure a2a, not a2av ?

**Comment by Sergei-Lebedev** [comment]
> > do you think we need similar algorithm for alltoall?
> 
> I think we can add if see a use case : does FB run pure a2a, not a2av ?

in pytorch we will run UCC alltoall in case of equal split https://github.com/facebookresearch/torch_ucc/blob/d29338636a968640228e6f2e6cbbf8eda288a07d/src/torch_ucc.cpp#L944.
IIRC param benchmark uses equal split for alltoall test

**Comment by vspetrov** [comment]
> > > do you think we need similar algorithm for alltoall?
> > 
> > 
> > I think we can add if see a use case : does FB run pure a2a, not a2av ?
> 
> in pytorch we will run UCC alltoall in case of equal split https://github.com/facebookresearch/torch_ucc/blob/d29338636a968640228e6f2e6cbbf8eda288a07d/src/torch_ucc.cpp#L944. IIRC param benchmark uses equal split for alltoall test

@Sergei-Lebedev Added split node alltoall to the PR

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**